File: Leetcode/labuladong_template/cpp-template/leetcode/editor/cn/binary-tree-preorder-traversal.cpp
#include <iostream>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    vector<int> res;

    // 返回前序遍历结果
    vector<int> preorderTraversal(TreeNode* root) {
        traverse(root);
        return res;
    }

    // 二叉树遍历函数
    void traverse(TreeNode* root) {
        if (root == nullptr) {
            return;
        }
        // 前序遍历位置
        res.push_back(root->val);
        traverse(root->left);
        traverse(root->right);
    }
};

//leetcode submit region end(Prohibit modification and deletion)


int main() {
    // your test code here
    Solution solution;
    TreeNode *root = TreeNode::createRoot({1, 22, 3, 4, 5, nullopt, 8, nullopt, nullopt, 6, 7, 9});
    vector<int> result = solution.preorderTraversal(root);
    for (int val : result) {
        cout << val << " ";
    }
}

File: Leetcode/labuladong_template/python-template/leetcode/editor/cn/BinaryTreePreorderTraversal.py
from idlelib.tree import TreeNode
from typing import *
from leetcode.editor.common.node import *


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # def __init__(self):
    #     self.res = []
    #
    # def traverse(self, root: Optional[TreeNode]):
    #     if root is None:
    #         return
    #
    #     self.res.append(root.val)
    #     self.traverse(root.left)
    #     self.traverse(root.right)
    #
    # def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
    #     self.traverse(root)
    #
    #     return self.res

    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        if root is None:
            return res

        res.append(root.val)

        res.extend(self.preorderTraversal(root.left))
        res.extend(self.preorderTraversal(root.right))

        return res

# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    solution = Solution()
    # your test code here